lib/phoenix/socket/transport.ex

```diff
@@ -118,17 +137,104 @@ defmodule Phoenix.Socket.Transport do
   implementation.
   """
 
-  require Logger
-  alias Phoenix.Socket
-  alias Phoenix.Socket.Message
-  alias Phoenix.Socket.Reply
+  @type state :: term()
 
-  @protocol_version "2.0.0"
+  @doc """
+  Returns a child specification for socket management.
+
+  It receives the endpoint configuration and is started when
+  the endpoint starts.
+  """
+  @callback child_spec(keyword) :: Supervisor.child_spec
+
+  @doc """
+  Connects to the socket.
+
+  The transport passes a map of metadata and the socket
+  returns `{:ok, state}` or `:error`. The state must be
+  stored by the transport and returned in all future
+  operations.
+
+  This function is used for authorization purposes and it
+  may be invoked outside of the process that effectively
+  runs the socket.
+
+  In the default `Phoenix.Socket` implementation, the
+  metadata expects the following keys:
+
+    * endpoint - the application endpoint
+    * serializer - the message serializer
+    * transport - the transport name
+    * params - the connection parameters
+
+  """
+  @callback connect(map) :: {:ok, state} | :error
 
   @doc """
-  Provides a keyword list of default configuration for socket transports.
+  Initializes the socket state.
+
+  This must be executed from the process that will effectively
+  operate the socket.
   """
-  @callback default_config() :: Keyword.t
+  @callback init(state) :: {:ok, state}
+
+  @doc """
+  Handles incoming socket messages.
+
+  The message is represented as `{payload, options}`. It must
+  return one of:
+
+    * `{:ok, state}` - continues the socket with no reply
+    * `{:reply, status, reply, state}` - continues the socket with reply
+    * `{:stop, reason, state}` - stops the socket
+
+  In the default `Phoenix.Socket` implementation, it calls the
+  `decode!` function in serializer with the given options which
+  must return a Phoenix.Socket message. The messages are usually
+  sent to the appropriate channel except:
+
+    * "heartbeat" events in the "phoenix" topic - should just emit
+      an OK reply
+    * "phx_join" on any topic - should join the topic
+    * "phx_leave" on any topic - should leave the topic
+
+  """
+  @callback handle_in({message :: term, opts :: keyword}, state) ::
+              {:ok, state}
+              | {:reply, status :: atom, {opcode :: atom, message :: term}, state}
+              | {:stop, reason :: term, state}
```

Comment:
Would it also be possible to support hibernation here and in `handle_info`?

Reply:
It would be possible - I don't plan to tackle it in this PR though.